iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/SnapshotDiffApplier.java

```diff
@@ -0,0 +1,422 @@
+package com.linkedin.openhouse.internal.catalog;
+
+import static com.linkedin.openhouse.internal.catalog.mapper.HouseTableSerdeUtils.getCanonicalFieldName;
+
+import com.linkedin.openhouse.cluster.metrics.micrometer.MetricsReporter;
+import com.linkedin.openhouse.internal.catalog.exception.InvalidIcebergSnapshotException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.SnapshotRef;
+import org.apache.iceberg.SnapshotSummary;
+import org.apache.iceberg.TableMetadata;
+
+/**
+ * Service responsible for applying snapshot changes to Iceberg table metadata.
+ *
+ * <p>The main entry point applySnapshots() has a clear flow: parse input → compute diff → validate
+ * → apply.
+ */
+@AllArgsConstructor
+@Slf4j
+public class SnapshotDiffApplier {
+
+  private final MetricsReporter metricsReporter;
+
+  /**
+   * Applies snapshot updates from metadata properties. Simple and clear: parse input, compute diff,
+   * validate, apply, record metrics, build.
+   *
+   * @param existingMetadata The existing table metadata (may be null for table creation)
+   * @param providedMetadata The new metadata with properties containing snapshot updates
+   * @return Updated metadata with snapshots applied
+   * @throws NullPointerException if providedMetadata is null
+   */
+  public TableMetadata applySnapshots(
+      TableMetadata existingMetadata, TableMetadata providedMetadata) {
+    // Validate at system boundary
+    Objects.requireNonNull(providedMetadata, "providedMetadata cannot be null");
+
+    String snapshotsJson = providedMetadata.properties().get(CatalogConstants.SNAPSHOTS_JSON_KEY);
+    if (snapshotsJson == null) {
+      return providedMetadata;
+    }
+
+    // Parse input
+    List<Snapshot> providedSnapshots = SnapshotsUtil.parseSnapshots(null, snapshotsJson);
+    Map<String, SnapshotRef> providedRefs =
+        Optional.ofNullable(providedMetadata.properties().get(CatalogConstants.SNAPSHOTS_REFS_KEY))
+            .map(SnapshotsUtil::parseSnapshotRefs)
+            .orElse(new HashMap<>());
```

Comment:
```suggestion
    String snapshotsJson = providedMetadata.properties().get(CatalogConstants.SNAPSHOTS_JSON_KEY);
    Map<String, SnapshotRef> providedRefs =
        Optional.ofNullable(providedMetadata.properties().get(CatalogConstants.SNAPSHOTS_REFS_KEY))
            .map(SnapshotsUtil::parseSnapshotRefs)
            .orElse(new HashMap<>());
    for (Map.Entry<String, SnapshotRef> entry : providedRefs.entrySet()) {
      if (!entry.getKey().equals(SnapshotRef.MAIN_BRANCH)) {
        throw new UnsupportedOperationException("OpenHouse supports only MAIN branch");
      }
    }        
    if (snapshotsJson == null) {
      return providedMetadata;
    }

    // Parse input
    List<Snapshot> providedSnapshots = SnapshotsUtil.parseSnapshots(null, snapshotsJson);
```

Can I suggest this for the refactor PR1? In PR2, this UnsupportedOperationException should go away, and we should explore if we need to look at both SNAPSHOTS_JSON_KEY and SNAPSHOTS_REFS_KEY before return the same providedMetadata?